File: Exercicio2/table.py
import math
 
class Table:
    def __init__(self, sequenciaInicial):
        self.dimensao = int(math.sqrt(len(sequenciaInicial)))
        self.matriz = [[0 for _ in range(self.dimensao)] for _ in range(self.dimensao)]
        k = 0

        for i in range(self.dimensao):
            for j in range(self.dimensao):
                self.matriz[i][j] = sequenciaInicial[k]
                if sequenciaInicial[k] == 0:
                    self.linha = i
                    self.coluna = j
                k += 1

    def movimento(self, movimento):
        linha = self.linha
        coluna = self.coluna

        if movimento == 'u' and linha != self.dimensao - 1:
            self.matriz[linha][coluna], self.matriz[linha + 1][coluna] = self.matriz[linha + 1][coluna], self.matriz[linha][coluna]
            self.linha = self.linha + 1

        elif movimento == 'd' and linha != 0:
            self.matriz[linha][coluna], self.matriz[linha - 1][coluna] = self.matriz[linha - 1][coluna], self.matriz[linha][coluna]
            self.linha = self.linha - 1

        elif movimento == 'l' and coluna != self.dimensao - 1:
            self.matriz[linha][coluna], self.matriz[linha][coluna + 1] = self.matriz[linha][coluna + 1], self.matriz[linha][coluna]
            self.coluna = self.coluna + 1

        elif movimento == 'r' and self.coluna != 0:
            self.matriz[linha][coluna], self.matriz[linha][coluna - 1] = self.matriz[linha][coluna - 1], self.matriz[linha][coluna]
            self.coluna = self.coluna - 1

    def printTable(self):
        index1 = 0
        index2 = 0

        for i in range(self.dimensao):
            for j in range(self.dimensao):
                print("+", end="")
                for k in range(6):
                    print("-", end="")
            print("+\n", end="")

            for index in range(self.dimensao):
                if self.matriz[index1][index2] >= 10:
                    print("|  ", end="")
                else:
                    print("|   ", end="")
                if self.matriz[index1][index2] == 0:
                    print(" ", end="")
                else:
                    print(self.matriz[index1][index2], end="")
                print("  ", end="")
                index2 += 1
            print("|\n", end="")
            index1 += 1
            index2 = 0

        for counter in range(self.dimensao):
            print("+", end="")
            for counter2 in range(6):
                print("-", end="")
        print("+\n\n", end="")

    def complete(self):
        verifica = True
        index = 0

        for i in range(self.dimensao):
            for j in range(self.dimensao):
                if self.matriz[i][j] != index:
                    verifica = False
                    break
                index = index + 1

        print("Posicao final: " + str(verifica))

File: Exercicio1/Table.java
import java.util.*;

public class Table {
    private int[][] table; //Matriz com números do tabuleiro
    private int row; // Linha do espaço vazio
    private int column; // Coluna do espaço vazio

    /*public Table() { // Construtor padrão
        over = false;
        table = new int[3][3];

        table[0][0] = 1;
        table[0][1] = 4;
        table[0][2] = 2;
        table[1][0] = 3;
        table[1][1] = 0;
        table[1][2] = 5;
        table[2][0] = 6;
        table[2][1] = 7;
        table[2][2] = 8;

        row = 1;
        column = 1;
    }*/

    public Table(ArrayList<Integer> numbers, int size) { //Construtor com posição predefinido 
        table = new int[size][size];
        int index = 0;

        for(int i = 0; i < table.length; i++) {
            for(int j = 0; j < table.length; j++) {
                table[i][j] = numbers.get(index++);
                if(table[i][j] == 0) {
                    row = i;
                    column = j;
                }
            }
        }
    }

    public void move(char operation) { //Realiza trocas na matriz conforme a operação
        switch (operation) {
            case 'u': //Troca entre 0 e seu valor abaixo
                if(row + 1 <= table.length - 1) {
                    int number = table[row+1][column];
                    table[row][column] = number;
                    table[row+1][column] = 0;
                    row++;
                }
                break;
            case 'l': //Troca entre 0 e seu valor à esquerda
                if(column + 1 <= table.length - 1) {
                    int number = table[row][column+1];
                    table[row][column] = number;
                    table[row][column+1] = 0;
                    column += 1;
                }
                break;
            case 'd': //Troca entre 0 e seu valor acima
                if(row - 1 >= 0) {
                    int number = table[row-1][column];
                    table[row][column] = number;
                    table[row-1][column] = 0;
                    row--;
                }
                break;
            case 'r': //Troca entre 0 e seu valor à direita
                if(column - 1 >= 0) {
                    int number = table[row][column-1];
                    table[row][column] = number;
                    table[row][column-1] = 0;
                    column -= 1;
                }
                break;
        }
    }

    public boolean complete() { //Verifica se os números estão em ordem crescente (completo)
        int number = 0;

        for(int i = 0; i < table.length; i++) {
            for(int j = 0; j < table.length; j++) {
                if(table[i][j] != number)
                    return false;
                number++;
            }
        }
        return true;
    }

    public void printTable() { //Imprime o tabuleiro conforme as formatações pedidas
        int index1, index2; //índices da matriz
        index1 = index2 = 0;

        for(int k = 0; k < table.length; k++) {
            for(int i = 0; i < table.length; i++) {
                System.out.print("+");
                for(int j = 0; j < 6; j++)
                    System.out.print("-");
            }
            System.out.println("+");
            
            for(int i = 0; i < table.length; i++) {
                if(table[index1][index2] >= 10)
                    System.out.print("|  ");
                else
                    System.out.print("|   ");
                System.out.print(table[index1][index2++] + "  ");
            }
            System.out.println("|");
            index1++;
            index2 = 0;
        }

        for(int k = 0; k < table.length; k++) {
            System.out.print("+");
            for(int i = 0; i < 6; i++)
                System.out.print("-");
        }
        System.out.println("+");
        System.out.println("");
    }
}